### PR TITLE
Implement the mimalloc option to speed up the compiler on Windows

### PR DESCRIPTION
**Pull request by hjyamauchi**
This is disabled by default.

### CODE REVIEW

**Comment by hjyamauchi** [comment]
@swift-ci please test

File: utils/build.ps1 [hunk 1]
```diff
@@ -102,6 +102,9 @@ in batch file format instead of executing them.
 .PARAMETER HostArchName
 The architecture where the toolchain will execute.
 
+.PARAMETER EnableMimalloc
```
Comment (compnerd) [inline utils/build.ps1:105]: I wonder if we should generalise this to something like `-allocator ...` to experiment with other allocators.

File: utils/build.ps1 [hunk 2]
```diff
@@ -1465,6 +1469,57 @@ function Build-Compilers() {
...
+# TODO: Add ARM64
+function Build-Mimalloc() {
+  [CmdletBinding(PositionalBinding = $false)]
+  param
+  (
+    [Parameter(Position = 0, Mandatory = $true)]
+    [hashtable]$Arch
+  )
+
+  if ($Arch -eq $ArchX64) {
```
Comment (compnerd) [inline utils/build.ps1:1482]: Can we abort if a different architecture is passed? Or is this being allowed for some other reason? As this currently stands, it feels like we could silently fail to enable something :(.

File: utils/build.ps1 (hunk 1 above)
Comment (hjyamauchi) [reply utils/build.ps1]: Done

File: utils/build.ps1 (hunk 2 above)
Comment (hjyamauchi) [reply utils/build.ps1:1484]: I think if we do that, the cross-compiling CI would likely fail

**Comment by hjyamauchi** [comment]
@swift-ci please test

File: utils/build.ps1 (hunk 2 above)
Comment (compnerd) [reply utils/build.ps1:1484]: Hmm, why would it fail? The check should be for what host we are compiling for, not what the build is right?

Comment (hjyamauchi) [reply utils/build.ps1:1484]: Yeah, I was probably thinking if the mimalloc flag were to be enabled by default.

Done.

**Comment by hjyamauchi** [comment]
@swift-ci please test

**Comment by hjyamauchi** [comment]
@swift-ci please test Linux Platform

File: utils/build.ps1 [hunk 3]
```diff
@@ -1465,6 +1471,59 @@ function Build-Compilers() {
...
+      $Args += "-i"
+      $Args += "-v"
+      $Args += $ExePath
+      Invoke-Program "$SourceCache\mimalloc\bin\minject" @Args
+      # Log the import table
+      $Args = @()
+      $Args += "-l"
+      $Args += $ExePath
+      Invoke-Program "$SourceCache\mimalloc\bin\minject" @Args
+      dir "$ExePath"
```
Comment (compnerd) [inline utils/build.ps1:1520]: Debugging left overs?

**Review by compnerd — APPROVED** [review]

File: utils/build.ps1 (hunk 3 above)
Comment (hjyamauchi) [reply utils/build.ps1:1520]: I mean to keep it so that we can verify that the import table is updated in the toolchain build log

**Comment by gregcotten** [comment]
@hjyamauchi have you observed significant compiler performance gains using mimalloc? Would you recommend the MS Swift Toolchain use it by default in an upcoming release?

**Comment by hjyamauchi** [comment]
@gregcotten In our internal app build, the build was about ~4% faster (which we think is meaningful in our long, long build times) and it seemed stable/robust. We have a plan to have it enabled by default in our future internal toolchain update but that's currently blocked by various other factors. I encourage testing it out independently, if you can :)

**Comment by gregcotten** [comment]
> @gregcotten In our internal app build, the build was about ~4% faster (which we think is meaningful in our long, long build times) and it seemed stable/robust. We have a plan to have it enabled by default in our future internal toolchain update but that's currently blocked by various other factors. I encourage testing it out independently, if you can :)

Thanks for the info! After adopting macros into a Windows Swift production app compile times have grown exponentially and I'm looking for any savings possible! Not sure if it's worth the hassle of maintaining my own toolchain though...

**Comment by hjyamauchi** [comment]
Sure! We, BCNY, have our own toolchain builds where it's enabled by default, if you're interested (note they're separate from the official swift.org builds): https://github.com/thebrowsercompany/swift-build

We may be able to enable it by default upstream but I cannot tell whether or when it would happen :)

**Pull request merged**